src/iter.rs

```diff
@@ -0,0 +1,685 @@
+
+        match trace {
+            Err(DecoderErrorInt::Eof) => None,
+            Err(DecoderErrorInt::Io(io)) => Some(Err(DecoderError::Io(io))),
+            Err(DecoderErrorInt::MalformedPacket(m)) => Some(Err(DecoderError::MalformedPacket(m))),
+            Ok(trace) => Some(Ok(trace)),
+        }
+    }
+}
+
+fn calc_offset(ts: u64, prescaler: Option<LocalTimestampOptions>, freq: u32) -> chrono::Duration {
+    let prescale = match prescaler {
+        None | Some(LocalTimestampOptions::Enabled) => 1,
+        Some(LocalTimestampOptions::EnabledDiv4) => 4,
+        Some(LocalTimestampOptions::EnabledDiv16) => 16,
+        Some(LocalTimestampOptions::EnabledDiv64) => 64,
+        Some(LocalTimestampOptions::Disabled) => unreachable!(), // checked in `Timestamps::new`
+    };
+    let ticks = ts * prescale;
+    let seconds = (1_f64 / freq as f64) * ticks as f64;
+
+    chrono::Duration::nanoseconds((seconds * 1e9).ceil() as i64)
+}
+
+#[cfg(test)]
+mod timestamp_utils {
+    use super::*;
+
+    #[test]
+    fn gts() {
+        let gts = Gts {
+            lower: None,
+            upper: None,
+        };
+        assert_eq!(gts.merge(), None, "noop merge");
+
+        let mut gts = Gts {
+            lower: Some(42),
+            upper: Some(42),
+        };
+        assert_eq!(
+            gts.merge(),
+            Some((42 << Gts::GTS2_SHIFT) | 42),
+            "(42, 42) merge"
+        );
+
+        gts.replace_lower(0b1101011);
+        assert_eq!(
+            gts.merge(),
+            Some((42 << Gts::GTS2_SHIFT) | 0b1101011),
+            "replace whole merge"
+        );
+
+        let mut gts = Gts {
+            lower: Some(42),
+            upper: Some(42),
+        };
+        gts.replace_lower(1);
+        assert_eq!(
+            gts.merge(),
+            Some((42 << Gts::GTS2_SHIFT) | 43),
+            "replace partial merge"
+        );
+    }
+
+    #[test]
+    fn offset() {
+        assert_eq!(
+            calc_offset(1000, Some(LocalTimestampOptions::EnabledDiv4), 16_000_000),
+            chrono::Duration::microseconds(250),
+        );
+    }
+}
+
+#[cfg(test)]
+mod timestamps {
+    use super::{calc_offset, Gts};
+    use crate::*;
+    use std::ops::{Add, Sub};
+
+    const FREQ: u32 = 16_000_000;
+    static mut BASELINE: Option<chrono::DateTime<chrono::Utc>> = None;
+
+    fn outer_calc_offset(
+        lts: TracePacket,
+        gts: Option<Gts>,
+        offset_sum: &mut chrono::Duration,
+    ) -> Timestamp {
+        let (reset, offset, data_relation) = match (lts, gts) {
+            (
+                TracePacket::LocalTimestamp1 {
+                    ts: lts,
+                    data_relation,
+                },
+                Some(gts),
+            ) => {
+                (
+                    true,
+                    calc_offset(gts.merge().unwrap() + lts, None, FREQ)
+                        // XXX why are we 1ns off?
+                        .sub(chrono::Duration::nanoseconds(1)),
```

Comment:
This unexplained nanosecond offset should be fixed before merge.

Reply:
Fixed. A simple off-by-one caused by passing in `gts + lts` instead of two seperate `calc_offset` calls.